sklearn/neighbors/_base.py

```diff
@@ -202,30 +202,65 @@ def _check_precomputed(X):
     graph = check_array(graph, accept_sparse="csr")
     check_non_negative(graph, whom="precomputed distance matrix.")
 
-    if not _is_sorted_by_data(graph):
+    if not _is_sorted_by_row_values(graph):
         warnings.warn(
-            "Precomputed sparse input was not sorted by data.", EfficiencyWarning
+            "Precomputed sparse input was not sorted by row values. Use the function "
+            "sklearn.neighbors.sort_by_row_values to sort the input by row values "
+            "and removes this warning.",
+            EfficiencyWarning,
         )
-        if not copied:
-            graph = graph.copy()
+        graph = sort_by_row_values(graph, copy=not copied)
+    return graph
 
-        # if each sample has the same number of provided neighbors
-        row_nnz = np.diff(graph.indptr)
-        if row_nnz.max() == row_nnz.min():
-            n_samples = graph.shape[0]
-            distances = graph.data.reshape(n_samples, -1)
 
-            order = np.argsort(distances, kind="mergesort")
-            order += np.arange(n_samples)[:, None] * row_nnz[0]
-            order = order.ravel()
-            graph.data = graph.data[order]
-            graph.indices = graph.indices[order]
+def sort_by_row_values(graph, copy=True):
+    """Sort a sparse graph such that each row is stored with increasing values.
+
+    Parameters
+    ----------
+    graph : sparse matrix, (n_samples, n_samples)
+        Distance matrix to other samples, where only non-zero elements are
+        considered neighbors. Matrix is converted to CSR format if not already.
+
+    copy : bool, optional (default=True)
+        If True, the graph is copied before sorting. If False, the sorting is
+        performed inplace. If graph is not of CSR format, a copy is always
+        returned.
```

Comment:
To reduce complexity of this function, I think we can remove `copy` and always sort in-place.

What do you think?